ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/handler/Handler.java

```diff
@@ -0,0 +1,179 @@
+
+  public Table resolveTable(String dbName, String tableName) throws MetaException {
+    try {
+      return getMSForConf(conf).getTable(getDefaultCatalog(conf), dbName, tableName);
+    } catch (MetaException e) {
+      LOG.error("Unable to find table {}.{}, {}", dbName, tableName, e.getMessage());
+      throw e;
+    }
+  }
+
+  protected Partition resolvePartition(String dbName, String tableName, String partName) throws MetaException {
+    if (partName != null) {
+      List<Partition> parts;
+      try {
+        parts = CompactorUtil.getPartitionsByNames(conf, dbName, tableName, partName);
+        if (parts == null || parts.isEmpty()) {
+          // The partition got dropped before we went looking for it.
+          return null;
+        }
+      } catch (Exception e) {
+        LOG.error("Unable to find partition: {}.{}.{}", dbName, tableName, partName, e);
+        throw e;
+      }
+      if (parts.size() != 1) {
+        LOG.error("{}.{}.{} does not refer to a single partition. {}", dbName, tableName, partName,
+                Arrays.toString(parts.toArray()));
+        throw new MetaException(String.join("Too many partitions for : ", dbName, tableName, partName));
+      }
+      return parts.get(0);
+    } else {
+      return null;
+    }
+  }
+
+  public <B extends TBase<B,?>> B computeIfAbsent(String key, Callable<B> callable) throws Exception {
+    if (metaCache.isPresent()) {
+      try {
+        return (B) metaCache.get().get(key, callable);
+      } catch (ExecutionException e) {
+        throw (Exception) e.getCause();
+      }
+    }
+    return callable.call();
+  }
+
+  Optional<Cache<String, TBase>> initializeCache(boolean tableCacheOn) {
+    if (tableCacheOn) {
+      metaCache = Optional.of(CacheBuilder.newBuilder().softValues().build());
+    }
+    return metaCache;
+  }
+
+  public void invalidateMetaCache() {
+    metaCache.ifPresent(Cache::invalidateAll);
+  }
+
+  protected List<Path> getObsoleteDirs(AcidDirectory dir, boolean isDynPartAbort) {
+    List<Path> obsoleteDirs = dir.getObsolete();
+    /*
+     * add anything in 'dir'  that only has data from aborted transactions - no one should be
+     * trying to read anything in that dir (except getAcidState() that only reads the name of
+     * this dir itself)
+     * So this may run ahead of {@link CompactionInfo#highestWriteId} but it's ok (suppose there
+     * are no active txns when cleaner runs).  The key is to not delete metadata about aborted
+     * txns with write IDs > {@link CompactionInfo#highestWriteId}.
+     * See {@link TxnStore#markCleaned(CompactionInfo)}
+     */
+    obsoleteDirs.addAll(dir.getAbortedDirectories());
+    if (isDynPartAbort) {
+      // In the event of an aborted DP operation, we should only consider the aborted directories for cleanup.
+      // Including obsolete directories for partitioned tables can result in data loss.
+      obsoleteDirs = dir.getAbortedDirectories();
+    }
+    return obsoleteDirs;
+  }
```

Comment:
Moved `getObsoleteDirs` to CompactorUtil since its using AcidDirectory which might not be generic to all handlers. However, the caching is done on a handler level now. Hence, caching functions are kept in handlers itself.
`resolvePartitions` seems generic to most handlers, hence kept it here.